### ISSUE TITLE
ipv4 validation not implemented

### ISSUE THREAD

**Issue by sirockin** [json-editor/json-editor#421]
### General information
The ipv4 format specifies that the entered string should be in the format of an ipv4 address. I believe this was already specified in json schema v4 but doesn't seem to be implemented in json-editor.

- json-editor version: 1.4.0-beta.0
### Expected behavior

Should create a validation error if a non-valid ip address is entered into a string field specified with format ipv4.

### Actual behavior

Does not raise a validation error.

### Steps to reproduce the behavior

[Direct link to example](https://json-editor.github.io/json-editor/?schema=N4IgLglmA2CmIC4QCsDOB7AdgAlQYwAtYBbAQ21gA9TiAHOEAGnAE9b4l0AjZWPMJiFoAndO2GRYqRKAi0AggBNFwqdISgwbDiFRhhETAHNBAM3TCyApHIBuAFkGKpeA7UhZEIAJIAFbKTKqqioAHQgAL4RzKoAjgCuEKqKiADaIHJKKmogALoRQA===&value=N4IglgDgggJjBOBTAzskAuEAjANgQwAsQBfIA===&lib_switcher=&prompt_before_delete&upload=function(a,b,c){console.log(%22Upload%20handler%20required%20for%20upload%20editor%22)}&theme=bootstrap2&iconlib=fontawesome4&object_layout=normal&show_errors=interaction)

```json
{
  "title": "json schema example",
  "type": "object",
  "properties": {
    "ipAddress": {
      "type": "string",
      "format": "ipv4",
      "description": "IP address."
    }
  },
  "required": [
    "ipAddress"
  ]
}
```

**Comment by sirockin** [comment]
I wrote a custom validator below that will provide this validation. Would you like me to attempt to create a PR to include make it part of the default validation?

```javascript
  JSONEditor.defaults.custom_validators.push(function(schema, value, path) {
    var errors = [];
    if(schema.format==="ipv4") 
    {
      try
      {
        let parts=value.split('.');
        if( parts.length !==4)
        {
          throw(Error());
        }
        parts.forEach((part)=>{
          if( isNaN(+part) || +part < 0 || +part>255)
          {
            throw(Error());
          }
        })
      }
      catch(err)
      {
        // Errors must be an object with `path`, `property`, and `message`
        errors.push({
          path: path,
          property: 'format',
          message: 'ip addresses must be in the form of 4 numbers between 0 and 255, separated by dots'
        });
      }
    }
    return errors;

```

**Comment by pmk65** [comment]
@sirockin Good idea, as both ipv4 and ipv6 is part of the 04 specs. https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-7.3.4
So the validation should support both versions. Also it might also be a good idea to create an "ip editor". It just has to extend the "string" editor and set a validator pattern, similar to what I did in the UUID editor `postBuild()` function.  https://github.com/json-editor/json-editor/blob/master/src/editors/uuid.js

**Comment by sirockin** [comment]
Thanks @pmk65 . 

So I'm guessing the right place to add the validation code would be in validator.js, just before the invocation of the custom validators. Is there any way I can include unit tests?

Regarding the "ip editor" idea, that sounds great but I'm not clear on what the UUID editor offers other than a default value. I can see how a box which enforces the numerical pattern (similar to the integer date editor) would be useful though. Do you have any simple examples of something like that?

**Comment by pmk65** [comment]
@sirockin I don't think an ip editor needs much more than a `postBuild()` function that sets the validation pattern. As you have noticed with the integer editor, then it doesn't need much code as it extends one of the base classes.
The UUID editor also has a feature to create a random UUID . That's why it contains more than just the `postBuild()` function with the pattern def. (line 10)
I believe that the way I set it, `this.schema.pattern` for live input and `this.jsoneditor.validator.schema.properties[this.key].pattern` for validation makes code in validator.js obsolete. 
As `this.jsoneditor.validator.schema.properties[this.key].pattern` forces the validator to use this pattern when validating.

**Comment by sirockin** [comment]
At this stage I think I'll limit myself to adding ipv4/6 validation to validator.js and look at digging into the editor code once I have a bit more understanding of your codebase and functionality. What are your thoughts on adding unit testing? - I see that you have e2e setup via codecept.

**Comment by pmk65** [comment]
@sirockin The tests are made by @germanbisurgi I don't have that much experience with setting up tests myself.
As for the codebase, then I have been working on refactoring several of the editors (and adding a couple new ones too) See #401 
Building new editors should be a bit easier as they now extend base classes. 😄

**Comment by schmunk42** [comment]
Unit tests would be also great, please feel free to add them if you have good ideas how they should be designed.

**Comment by pmk65** [comment]
@sirockin Here's an example of an "ip editor", like I described before: https://is.gd/MZKWaq

**Comment by sirockin** [comment]
@schmunk42 Okay great, I've added #423 and will provide a PR for that.

**Comment by pmk65** [comment]
I added a PR #425 to allow setting alternate pattern validation message.

**Comment by sirockin** [comment]
I have an ipv4 and ipv6 implementation and unit tests ready to submit as a Pr once my previous PR has built and been merged. The regex pattern supplied above for ipv4 doesn't catch out-of-range numbers so I've stuck to my implementation. The ipv6 pattern looks pretty good although it doesn't allow compressed-form addresses. However due to lack of time I've decide to use it and have left in a disabled unit test to allow someone else to provide this when convenient.

**Comment by sirockin** [comment]
@pmk65 if #426 covers everything that's great - I'll be happy for you to reject #428 if that's the case. One thing I would ask though - can we try to adapt the unit tests in my PR to cover your implementation?

**Comment by pmk65** [comment]
@sirockin Sure. The unit tests would be great to have. Maybe you can add them to #426 or create a companion PR. 🥇
(As I said before, I have very little experience in setting up tests myself. 🙈)

**Comment by schmunk42** [comment]
fixed by #426
